trinity/db/orm.py

```
@@ -0,0 +1,113 @@
+from pathlib import Path
+
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.exc import (
+    OperationalError,
+)
+from sqlalchemy.orm.exc import (
+    NoResultFound,
+    MultipleResultsFound,
+)
+from sqlalchemy.orm import (
+    sessionmaker,
+    Session as BaseSession,
+)
+from sqlalchemy import (
+    create_engine,
+    Column,
+    Integer,
+    String,
+)
+
+from p2p.exceptions import (
+    BadDatabaseError,
+)
+
+
+Base = declarative_base()
+
+
+SCHEMA_VERSION = '2'
+
+
+class SchemaVersion(Base):
+    __tablename__ = 'schema_version'
+
+    id = Column(Integer, primary_key=True)
+    version = Column(String, unique=True, nullable=False, index=True)
+
+
+#
+# SQL Based Trackers
+#
+def _get_session(path: Path) -> 'BaseSession':
+    # python 3.6 does not support sqlite3.connect(Path)
+    is_memory = path.name == ':memory:'
+
+    if is_memory:
+        database_uri = 'sqlite:///:memory:'
+    else:
+        database_uri = f'sqlite:///{path.resolve()}'
+
+    engine = create_engine(database_uri)
+    Session = sessionmaker(bind=engine)
+    session = Session()
+    return session
+
+
+def _setup_schema(session: BaseSession) -> None:
+    Base.metadata.create_all(session.get_bind())
+    session.add(SchemaVersion(version=SCHEMA_VERSION))
+    session.commit()
+
+
+def _get_schema_version(session: BaseSession) -> str:
+    schema_version = session.query(SchemaVersion).one()
+    return schema_version.version
+
+
+def _check_is_empty(session: BaseSession) -> bool:
+    engine = session.get_bind()
+    for table_name in Base.metadata.tables.keys():
```

I can't decide if this is dangerous or not but it makes me a little nervous. This only finds `BlacklistRecord` because `trinity.__init__` imports `main` which eventually imports the plugins. If any part of that chain is broken (or if someone later adds a model in a file which isn't automatically imported) then this method (and `_check_tables_exist`) will silently fail to look for the missing tables. I think this is unlikely but it's a weird enough error that if it does happen it'll take quite a while to debug, so it might be worth fixing now.

I think there are two options:
- bringing `BlacklistRecord` into this file, or some nearby file which is obviously where new models should be added, in much the same way `trinity.exceptions` is obviously where new exceptions should be added.
- Instead of looking for tables in the base, looking for tables in the engine. `_check_is_empty` becomes `len(session.bind.table_names()) == 0`. I'm not sure what to do about `_check_tables_exist` because you really want to look for *our* tables, so maybe the above answer is the better one.

---

I had the same issue.  Now that this lives in Trinity, I think I'm more comfortable just making this code be very explicity about which models it expects to be in place.